src/view/com/composer/char-progress/CharProgress.tsx

```diff
@@ -27,35 +26,51 @@ export function CharProgress({
   const pal = usePalette('default')
   const textColor = count > maxLength ? '#e60000' : pal.colors.text
   const circleColor = count > maxLength ? '#e60000' : pal.colors.link
+  const remainingCount = maxLength - count
+  const shouldShowText = remainingCount <= 20
+  const shouldShowCircle = remainingCount > -10
+
   return (
     <View
-      style={[a.flex_row, a.align_center, a.justify_between, a.gap_sm, style]}>
-      <Text
-        style={[
-          {color: textColor, fontVariant: ['tabular-nums']},
-          a.flex_grow,
-          a.text_right,
-          textStyle,
-        ]}>
-        {maxLength - count}
-      </Text>
-      {count > maxLength ? (
-        <ProgressPie
-          size={size ?? 30}
-          borderWidth={4}
-          borderColor={circleColor}
-          color={circleColor}
-          progress={Math.min((count - maxLength) / maxLength, 1)}
-        />
-      ) : (
-        <ProgressCircle
-          size={size ?? 30}
-          borderWidth={1}
-          borderColor={pal.colors.border}
-          color={circleColor}
-          progress={count / maxLength}
-        />
+      style={[
+        a.flex,
+        a.align_center,
+        a.justify_center,
+        a.gap_sm,
+        styles.container,
+        style,
+      ]}>
+      {shouldShowCircle && (
+        <Animated.View
+          exiting={FadeOut.duration(200)}
+          entering={FadeIn.duration(200)}
+          style={styles.progressCircle}>
+          <ProgressCircle
+            animated={false}
+            size={size ?? 30}
+            borderWidth={0}
+            unfilledColor={pal.colors.border}
+            color={circleColor}
+            progress={Math.min(count / maxLength, 1)}
+          />
+        </Animated.View>
+      )}
+      {shouldShowText && (
+        <Text
+          style={[{color: textColor, fontSize: 14}, a.text_right, textStyle]}>
+          {remainingCount}
+        </Text>
       )}
     </View>
   )
 }
+
+const styles = StyleSheet.create({
+  container: {
+    minWidth: 30,
+    height: 30,
+  },
+  progressCircle: {
+    position: 'absolute',
+  },
+})
```

Comment:
for readability I'd put the `styles` declaration above the component, instead of below it.